File: app/src/main/java/com/example/shenjack/zhihudailyreader/data/source/StoryRepository.java
package com.example.shenjack.zhihudailyreader.data.source;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.example.shenjack.zhihudailyreader.data.BeforePosts;
import com.example.shenjack.zhihudailyreader.data.Detail;
import com.example.shenjack.zhihudailyreader.data.Post;
import com.example.shenjack.zhihudailyreader.data.StoriesBean;
import com.example.shenjack.zhihudailyreader.data.TodayPosts;
import com.example.shenjack.zhihudailyreader.data.source.local.StoryLocalDataSource;
import com.example.shenjack.zhihudailyreader.data.source.remote.StoryRemoteDataSource;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static com.example.shenjack.zhihudailyreader.Util.Util.checkNotNull;


/**
 * Created by ShenJack on 2017/6/1.
 */

public class StoryRepository implements StoryDataSource {

    private static StoryRepository instance = null;

    private final StoryDataSource mStoryLocalDataSource;

    private final StoryDataSource mStoryRemoteDataSource;

    private List<PostRepositoryObserver> mObservers = new ArrayList<PostRepositoryObserver>();

    Map<String, Post> mCachedPosts;

    public List<Post> getCachedPosts(){
        return mCachedPosts == null? null: new ArrayList<>(mCachedPosts.values());
    }

    public Post getCachedPost(String postId){
        return mCachedPosts.get(postId);
    }

    private void saveTaskInLocalDataSource(List<StoriesBean> posts){
        if(posts !=null){
            for (StoriesBean post:posts
                    ){
                mStoryLocalDataSource.savePost(post);
            }
        }
    }

    private void processLoadedTasks(List<Post> posts){
        // TODO: 2017/6/2 not yet
    }


    public static StoryRepository getInstance(StoryLocalDataSource storyLocalDataSource,
                                              StoryRemoteDataSource postRemoteDataSource){
        if(instance == null){
            instance = new StoryRepository(storyLocalDataSource,postRemoteDataSource);
        }

        return instance;
    }

    public static StoryRepository getInstance(){
        if(instance == null){
            instance = new StoryRepository(StoryLocalDataSource.getInstance(), StoryRemoteDataSource.getInstance());
        }

        return instance;
    }

    public StoryRepository(StoryDataSource postLocalDataSource, StoryDataSource postRemoteDataSource) {
        mStoryLocalDataSource = checkNotNull(postLocalDataSource);
        mStoryRemoteDataSource = checkNotNull(postRemoteDataSource);
    }



    @Nullable
    @Override
    public io.reactivex.Observable<TodayPosts> getTodayStories() {
        return mStoryRemoteDataSource.getTodayStories();
    }

    @Override
    public io.reactivex.Observable<BeforePosts> getBeforeStories(String date) {
        return mStoryRemoteDataSource.getBeforeStories(date);
    }

    @Override
    public io.reactivex.Observable<List<TodayPosts.TopStoriesBean>> getTopStories() {
        return mStoryRemoteDataSource.getTopStories();
    }

    @Nullable
    @Override
    public io.reactivex.Observable<Detail> getStoryDetail(@NonNull String storyId) {
        return mStoryRemoteDataSource.getStoryDetail(storyId);
    }

    @Override
    public void savePost(@NonNull StoriesBean storiesBean) {

    }

    @Override
    public void readPost(@NonNull StoriesBean storiesBean) {

    }

    @Override
    public void readPost(@NonNull int postId) {

    }


    @Override
    public void refreshPosts() {

    }

    public interface PostRepositoryObserver {

        void onPostChanged();

    }
}

File: app/src/main/java/com/example/shenjack/zhihudailyreader/data/source/StoryRepository.kt
package com.example.shenjack.zhihudailyreader.data.source

import com.example.shenjack.zhihudailyreader.data.BeforeStories
import com.example.shenjack.zhihudailyreader.data.Detail
import com.example.shenjack.zhihudailyreader.data.StoriesBean
import com.example.shenjack.zhihudailyreader.data.TodayStories
import com.example.shenjack.zhihudailyreader.data.source.local.StoryLocalDataSource
import com.example.shenjack.zhihudailyreader.data.source.remote.StoryRemoteDataSource

import java.util.ArrayList

import com.example.shenjack.zhihudailyreader.Util.Util.checkNotNull
import io.reactivex.Observable


/**
 * Created by ShenJack on 2017/6/1.
 */

class StoryRepository(postLocalDataSource: StoryDataSource?, postRemoteDataSource: StoryDataSource?) : StoryDataSource {
    override val topStories: Observable<List<TodayStories.TopStoriesBean>>
        get() = mStoryRemoteDataSource.topStories!! //To change initializer of created properties use File | Settings | File Templates.
    override val todayStories: Observable<TodayStories>
        get() = mStoryRemoteDataSource.todayStories!! //To change initializer of created properties use File | Settings | File Templates.

    private val mStoryLocalDataSource: StoryDataSource

    private val mStoryRemoteDataSource: StoryDataSource

    init {
        mStoryLocalDataSource = postLocalDataSource!!
        mStoryRemoteDataSource = postRemoteDataSource!!
    }




    override fun getBeforeStories(date: String): io.reactivex.Observable<BeforeStories>? {
        return mStoryRemoteDataSource.getBeforeStories(date)
    }


    override fun getStoryDetail(storyId: String): io.reactivex.Observable<Detail>? {
        return mStoryRemoteDataSource.getStoryDetail(storyId)
    }




    interface PostRepositoryObserver {

        fun onPostChanged()

    }

    companion object {

        var instance: StoryRepository? = null
        get() {
            if(field == null) instance = StoryRepository(StoryLocalDataSource.instance,StoryRemoteDataSource.instance)
            return field
        }


        fun getInstance(storyLocalDataSource: StoryLocalDataSource,
                        postRemoteDataSource: StoryRemoteDataSource): StoryRepository? {
            if (instance == null) {
                instance = StoryRepository(storyLocalDataSource, postRemoteDataSource)
            }

            return instance
        }

    }

}
